lib/createExplorer.js

```diff
@@ -32,56 +32,88 @@ module.exports = function (options) {
       if (fileCache && fileCache.has(absoluteConfigPath)) {
         return fileCache.get(absoluteConfigPath);
       }
-      var result = loadDefinedFile(absoluteConfigPath, options)
-        .then(transform);
+      var result = (!options.sync)
+        ? loadDefinedFile(absoluteConfigPath, options)
+            .then(transform)
+        :  transform(loadDefinedFile(absoluteConfigPath, options));
       if (fileCache) fileCache.set(absoluteConfigPath, result);
       return result;
     }
 
-    if (!searchPath) return Promise.resolve(null);
+    if (!searchPath) return (!options.sync) ? Promise.resolve(null) : null;
 
     var absoluteSearchPath = path.resolve(process.cwd(), searchPath);
 
-    return isDirectory(absoluteSearchPath)
-      .then(function (searchPathIsDirectory) {
-        var directory = (searchPathIsDirectory)
+    return (!options.sync)
+      ? isDirectory(absoluteSearchPath)
+        .then(function (searchPathIsDirectory) {
+          var directory = (searchPathIsDirectory)
+            ? absoluteSearchPath
+            : path.dirname(absoluteSearchPath);
+          return searchDirectory(directory);
+        })
+      : searchDirectory(
+          isDir.sync(absoluteSearchPath)
           ? absoluteSearchPath
-          : path.dirname(absoluteSearchPath);
-        return searchDirectory(directory);
-      });
+          : path.dirname(absoluteSearchPath)
+        );
   }
 
   function searchDirectory(directory) {
     if (directoryCache && directoryCache.has(directory)) {
       return directoryCache.get(directory);
     }
 
-    var result = Promise.resolve()
-      .then(function () {
-        if (!options.packageProp) return;
-        return loadPackageProp(directory, options);
-      })
-      .then(function (result) {
-        if (result || !options.rc) return result;
-        return loadRc(path.join(directory, options.rc), options);
-      })
-      .then(function (result) {
-        if (result || !options.js) return result;
-        return loadJs(path.join(directory, options.js));
-      })
-      .then(function (result) {
-        if (result) return result;
-
+    var result;
+    if (!options.sync) {
+      result = Promise.resolve()
+        .then(function () {
+          if (!options.packageProp) return;
+          return loadPackageProp(directory, options);
+        })
+        .then(function (result) {
+          if (result || !options.rc) return result;
+          return loadRc(path.join(directory, options.rc), options);
+        })
+        .then(function (result) {
+          if (result || !options.js) return result;
+          return loadJs(path.join(directory, options.js), options);
+        })
+        .then(function (result) {
+          if (result) return result;
+
+          var splitPath = directory.split(path.sep);
+          var nextDirectory = (splitPath.length > 1)
+            ? splitPath.slice(0, -1).join(path.sep)
+            : null;
```

Comment:
I want to avoid duplicating lines like this here and on lines 106+. Can we make a function for it?